src/backend/drm/atomic/session.rs

```diff
@@ -95,5 +97,63 @@ impl<A: AsRawFd + 'static> SessionObserver for AtomicDrmDeviceObserver<A> {
             }
         }
         self.active.store(true, Ordering::SeqCst);
+        // okay, the previous session/whatever might left the drm devices in any state...
+        // lets fix that
+        if let Err(err) = self.reset_state() {
+            warn!(self.logger, "Unable to reset state after tty switch: {}", err);
+            // TODO call drm-handler::error
+        }
+    }
+}
+
+impl<A: AsRawFd + 'static> AtomicDrmDeviceObserver<A> {
+    fn reset_state(&mut self) -> Result<(), Error> {
+        if let Some(dev) = self.dev.upgrade() {
+            let res_handles = ControlDevice::resource_handles(&*dev)
+                .compat()
+                .map_err(|source| Error::Access {
+                    errmsg: "Error loading drm resources",
+                    dev: dev.dev_path(),
+                    source
+                })?;
+
+            // Disable all connectors (otherwise we might run into conflicting commits when restarting the rendering loop)
```

Comment:
I'm not sure I get it. The Legacy `reset_state` seems to restore the state to match whatever it was set before Smithay was previously disabled, while this atomic `reset_state` seems to just disable everything?

Reply:
Yup, because our *atomic* commits are *not atomic* enough (but are per surface).

First a clarification:
legacy's `reset_state` function does not restore the state, but captures the current state (which shows the previous tty or getty) and sets that into the *current* state variable of each surface.
We just synchronize the real state with our cached state, so we can correctly *diff* them on the next commit (and apply our desired config). Otherwise our surfaces would assume nothing has changed and fail.

Now for an example to show, why we do not use the same strategy for the atomic backend (but rather disable everything):

Lets say another tty switched out CRTCs around, so our *pending* state is:
- `crtc A -> connector 1` and `crtc B -> connector 2`
but the other tty changed this to: 
- `crtc B -> connector 1` and `crtc A -> connector 1`.

Using the legacy API the following happens:
- either crtc is first modeset (or maybe even parallel once `Surface` is `Send`), lets say this is A.
- `crtc A` submits a frame to `connector 1`  (because that is our state).
   - The kernel acknowledges this request, by disconnecting `connector 1` from `crtc B`
   - This works in legacy mode-setting. You can basically steal connectors.
- The surface of `crtc A` maybe null-commits thus deactivating `connector 2`, before the next step.
    - This causes flicker, but does nothing, if the next step happens earlier, which is why this is safe (important for the atomic case later)
- `crtc B` submits a frame to `connector 2` (stealing the connector, if there was no null-commit yet)
    - our state is restored at this point, no matter what the previous state was -> this works
    - The next null commit does nothing to our two connectors and the rest is disabled

Using the atomic api now the following *would* happen, if we employ the same strategy:
- Lets say we start with A again.
    - There is no null-commit in the atomic api, so `crtc A` commits `connector 1` is active and serves a frame *and* disables `connector 2` (because that is in the *current* state) (and we want to disable unused connectors).
    - The kernel rejects this, because `conn 1` is used by `crtc B` and our request says nothing about the state of `crtc B`, which would have no remaining connector, which is invalid.
    - To clarify: connector stealing is possible in atomic, but not if the resulting state would disable a crtc without explicitly setting that.
    - So this request would need to set `crtc B` to inactive (or into another valid config), but the Surface knows nothing about other surfaces and we do not want to synchronize commits of different crtcs (because they might run at different refresh rates).
- So we have a first fail, but it gets worse, `crtc B` fails for the same reason, because the commit is atomic, means `connector 2` was never disconnected from `crtc A` as the whole request was rejected (no partial state changes like with the legacy api).
     - As a result the user needs to reinitialize the whole device or disconnect all connectors before commiting again...

So our solution to this problem is an easy work-around:
We just disable everything (like on device creation) and the next commit applies our desired config again. Like this we cannot run into a situation, where we cannot enable a crtc, if we previously had a valid config (which we had).

The better fix would be to combine with first requests of all Surfaces, when we re-activate the session. But we would need to delay that until we have new frames for all of them or cache the old framebuffers, etc. I want to do this at some point, but it gets complicated enough to fix this later. All I want for now is a working configuration.